### PR TITLE
Added the ability to output a binary log from MSBuild.

### LINKED ISSUE #3625: MSBuild /bl flag for Binary Logging
MSBuild has a command line flag to activate the output of a Binary Log that can replay builds to gather diagnostics on builds.

I would be really useful to add this utility to the MSBuild helper.

[MSBuild Binary Log Docs](https://github.com/Microsoft/msbuild/blob/master/documentation/wiki/Binary-Log.md)

### PR DESCRIPTION
**Pull request by luckielordie**
This is a change that adds the MSBuild binary logging ability referenced by the issue #3625
Closes #3625

Changelog: Feature: Added parameter to the ``build()`` method of the ``MSBuild()`` build helper to create a binary log.

### CODE REVIEW

**Comment by jgsogo** [comment]
Hi! Thanks for taking this into account.

I would consider passing the output filename (https://github.com/Microsoft/msbuild/blob/master/documentation/wiki/Binary-Log.md#creating-a-binary-log-during-a-build) in the param ``output_binary_log``, not very fan of this semantics, but something like this can work:
- ``if not output_binary_log``, then do not add the ``/bl`` argument.
- ``if ouptut_binary_log and isinstance(output_binary_log, bool)``, then add just ``/bl``
- ``if output_binary_log``, then add ``"/bl:{}".format(output_binary_log)``

Also, it should be conditionally added to MSBuild >=15.3, whick is the first version to support it.

**Comment by luckielordie** [comment]
> Hi! Thanks for taking this into account.
> 
> I would consider passing the output filename (https://github.com/Microsoft/msbuild/blob/master/documentation/wiki/Binary-Log.md#creating-a-binary-log-during-a-build) in the param `output_binary_log`, not very fan of this semantics, but something like this can work:
> 
> * `if not output_binary_log`, then do not add the `/bl` argument.
> * `if ouptut_binary_log and isinstance(output_binary_log, bool)`, then add just `/bl`
> * `if output_binary_log`, then add `"/bl:{}".format(output_binary_log)`
> 
> Also, it should be conditionally added to MSBuild >=15.3, whick is the first version to support it.

All good suggestions! I'll try to update this tonight.

**Comment by luckielordie** [comment]
Sorry, in haste I clicked the wrong button.

**Comment by luckielordie** [comment]
What sort of behaviour is appropriate if the version does not support the binary logging feature? 

It's not fatal but do we still want the build to be completed? Should we warn and carry on or warn and fail?

Is there a mechanism in place already to get the msbuild version we're currently using?

**Comment by jgsogo** [comment]
> What sort of behaviour is appropriate if the version does not support the binary logging feature?
It's not fatal but do we still want the build to be completed? Should we warn and carry on or warn and fail?

I'm not sure. I think that this feature won't be enabled by default in a recipe, so you are going to use it when you are working on a recipe: you will activate it explicitly, look for that file, and use it for debugging. If this is the use case, I will raise.

> Is there a mechanism in place already to get the msbuild version we're currently using?

Other tools like CMake or SVN are adding an staticmethod called ``get_version`` to get this kind of information, you will need to create your own. Maybe in the future we need to unify them all, but until now there is no work in that direction.

https://github.com/conan-io/conan/blob/acd821c27a8d502c4919d6a70c289f97fc1d6d2b/conans/client/build/cmake.py#L329

**Comment by jgsogo** [comment]
@luckielordie, we need some changes in this PR to make it pass the test-suite.

And also, we need Matthew Lord to sign the CLA... can you tell him? it looks like he has no github account 😕

**Comment by luckielordie** [comment]
> @luckielordie, we need some changes in this PR to make it pass the test-suite.
> 
> And also, we need Matthew Lord to sign the CLA... can you tell him? it looks like he has no github account

That's me, my work email isn't attatched to my github, didn't realise at the time. 

I keep meaning to come around and fix this whenever I have time.

**Comment by jgsogo** [comment]
Just a kindly reminder... 😇

File: conans/client/build/msbuild.py [hunk 1]
```diff
@@ -87,6 +87,13 @@ def get_command(self, project_file, props_file_path=None, targets=None, upgrade_
                 self._output.warn("***** The configuration %s does not exist in this solution *****" % config)
                 self._output.warn("Use 'platforms' argument to define your architectures")
 
+        if output_binary_log not None:
```
Comment (memsharded) [inline conans/client/build/msbuild.py]: Typical comparison would be ``if output_binary_log is not None``. Seems an error

File: conans/client/build/msbuild.py [hunk 2]
```diff
@@ -87,6 +87,13 @@ def get_command(self, project_file, props_file_path=None, targets=None, upgrade_
...
+            if isinstance(output_binary_log, bool):
+                if output_binary_log:
+                    command.append(' /bl ')
```
Comment (memsharded) [inline conans/client/build/msbuild.py]: The join already takes into account whitespace, not necessary that you add extra ones here.

File: conans/client/build/msbuild.py [hunk 3]
```diff
@@ -87,6 +87,13 @@ def get_command(self, project_file, props_file_path=None, targets=None, upgrade_
...
+            else:
+                command.append(' /bl:"%s" '.format(output_binary_log))
```
Comment (memsharded) [inline conans/client/build/msbuild.py]: Either use ``"mystring %s" % datastring`` or ``"mystring {}".format()``

**Comment by danimtb** [comment]
Hi @luckielordie  👋 

We are willing to include this PR on the next Conan release. We could work over this same PR if you don't have time, but we need you to sign the CL to add the changes. Please update us and we will proceed accordingly.

Thanks! 😄

**Comment by luckielordie** [comment]
> Hi @luckielordie  
> 
> We are willing to include this PR on the next Conan release. We could work over this same PR if you don't have time, but we need you to sign the CL to add the changes. Please update us and we will proceed accordingly.
> 
> Thanks!

Sorry, let me sort this out on my lunch break. I'll get the CL signed and you're more than welcome to help me on it, I'm so short on time lately.

File: conans/test/util/build_sln_command_test.py [hunk 4]
```diff
@@ -144,3 +144,51 @@ def properties_file_test(self):
...
+        command = build_sln_command(MockSettings({"compiler": "Visual Studio",
+                                                  "compiler.version": "17",
+                                                  "build_type": "Debug",
+                                                  "compiler.runtime": "MDd",
+                                                  "cppstd": "17"}),
+                                    sln_path='dummy.sln', targets=None,
+                                    upgrade_project=False, build_type='Debug', arch='armv7',
+                                    parallel=False, output_binary_log=bl_filename)
+        
+        expected_command = '/bl:"%s"'.format(bl_filename)
```
Comment (danimtb) [inline conans/test/util/build_sln_command_test.py]: same here

File: conans/client/build/msbuild.py [hunk 5]
```diff
@@ -87,6 +89,15 @@ def get_command(self, project_file, props_file_path=None, targets=None, upgrade_
                 self._output.warn("***** The configuration %s does not exist in this solution *****" % config)
                 self._output.warn("Use 'platforms' argument to define your architectures")
 
+        if output_binary_log is not None and output_binary_log:
```
Comment (jgsogo) [inline conans/client/build/msbuild.py]:
```suggestion
        if output_binary_log:
```

File: conans/test/build_helpers/msbuild_test.py [hunk 6]
```diff
@@ -150,3 +154,114 @@ def build(self):
...
+    exports = "*"
+    settings = "os", "build_type", "arch", "compiler", "cppstd"
+
+    def build(self):
+        msbuild = MSBuild(self)
+        msbuild.build("MyProject.sln", output_binary_log=%s)
+"""
+        client = TestClient()
+        files = get_vs_project_files()
+        for value in ("True", "'my_log.binlog'"):
```
Comment (jgsogo) [inline conans/test/build_helpers/msbuild_test.py]: As they are two different tests, it is better to use the `parameterized` decorator for the test:

```
@parameterized.expand([("True", ), ("'my_log.binlog'", ) ])
```

File: conans/test/build_helpers/msbuild_test.py [hunk 7]
```diff
@@ -150,3 +154,114 @@ def build(self):
...
+            files[CONANFILE] = conan_build_vs % value
+            client.save(files)
+            client.run("install .")
+            client.run("build .")
+
+            if value == "'my_log.binlog'":
+                log_name = "my_log.binlog"
+                flag = "/bl:my_log.binlog"
```
Comment (jgsogo) [inline conans/test/build_helpers/msbuild_test.py]: We can use the variable `value` a little bit here.

File: conans/client/build/msbuild.py [hunk 8]
```diff
@@ -87,6 +89,15 @@ def get_command(self, project_file, props_file_path=None, targets=None, upgrade_
...
+            msbuild_version = MSBuild.get_version(self._settings)
+            if msbuild_version > "15.3":  # http://msbuildlog.com/
```
Comment (jgsogo) [inline conans/client/build/msbuild.py]:
```suggestion
            if msbuild_version >= "15.3":  # http://msbuildlog.com/
```

I understand that version 15.3 will work.

File: conans/test/build_helpers/msbuild_test.py [hunk 9]
```diff
@@ -150,3 +155,114 @@ def build(self):
...
+
+class HelloConan(ConanFile):
+    name = "Hello"
+    version = "1.2.1"
+    exports = "*"
+    settings = "os", "build_type", "arch", "compiler", "cppstd"
+
+    def build(self):
+        msbuild = MSBuild(self)
+        msbuild.build("MyProject.sln", output_binary_log="%s")
```
Comment (danimtb) [inline conans/test/build_helpers/msbuild_test.py]: @jgsogo Setting this to ``"%s"`` and formatting with value ``True`` will result in ``output_binary_log="True"`` which is not the desired result and now the test fails...

Comment (jgsogo) [reply conans/test/build_helpers/msbuild_test.py]: Oh! Sorry, this is a little bit embarrasing :sweat: I'll fix it myself.

**Pull request merged**